phase4-lib/src/main/java/com/helger/phase4/servlet/AS4IncomingHandler.java

```diff
@@ -977,16 +981,23 @@ public static Ebms3UserMessage parseUserMessage (@Nonnull final IAS4CryptoFactor
     final Ebms3UserMessage ret = aState.getEbmsUserMessage ();
     if (ret == null)
     {
-      if (aState.getEbmsSignalMessage () != null)
+      if (aState.getEbmsSignalMessage () != null) {
         LOGGER.warn ("A Message state is present, but it contains a SignalMessage instead of a UserMessage.");
-      else
-        LOGGER.warn ("A Message state is present, but it contains neither a SignalMessage nor a UserMessage.");
+        aSignalMsgConsumer.handleSignalMessage(aState.getEbmsSignalMessage(), aMessageMetadata, aState);
+      } else {
+        LOGGER.warn("A Message state is present, but it contains neither a SignalMessage nor a UserMessage.");
+      }
     }
     else
     {
       // Invoke consumer here, because we have the state
       if (aUserMsgConsumer != null)
         aUserMsgConsumer.handleUserMessage (ret, aMessageMetadata, aState);
+
+      if (aAttachmentConsumer != null) {

```

Comment:
Why is that block needed? The attachments can be extracted from the `aState` anyway and are therefore accessible in `handleUserMessage` anyway.... I don't understand the necessity.

Reply:
The Problem is, at the point where the user gets the `aState` the streams for the attachments are already closed and can not be opened again. I think this happens in the AS4IncomingHandler. This was my solution to get the attachments streams in a point where the data is still accessible